### ISSUE TITLE
ICE broken MIR when updating to 1.56.0 and edition 2021

### ISSUE THREAD

**Issue by rasviitanen** [rust-lang/rust#90158]
We just migrated to 1.56.0 and edition 2021 (from 1.53.0). This resulted in some internal compiler errors when using a let binding for an enum with a single variant (`let Enum::Variant(it) = it`).

This issue is not present in `rustc 1.53.0`.

I can unfortunately not share our repository and I have not been able to recreate this error elsewhere. I will do my best to provide additional information if needed.

Summary:
```
warning: Error finalizing incremental compilation session directory
error: internal compiler error: broken MIR in DefId ... can't project out of PlaceTy ...
error: TyKind::Error constructed but no error reported
```

More in the logs below.

### Code

```Rust
impl ContentScope {
    #[instrument(skip(global_state))]
    pub async fn has_permission(self, client: Client, global_state: &GlobalState) -> Result<()> {
        // This doesn't ICE
        // match self {
        //     ContentScope::Group(group_ref) => {
        //         client.check_permission(group_ref, global_state).await
        //     }
        // }

        let ContentScope::Group(group_ref) = self;
        client.check_permission(group_ref, global_state).await
    }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0 (09c42c458 2021-10-18)
binary: rustc
commit-hash: 09c42c45858d5f3aedfa670698275303a3d19afa
commit-date: 2021-10-18
host: x86_64-pc-windows-msvc
release: 1.56.0
LLVM version: 13.0.0
```

### Error output

```rust
warning: Error finalizing incremental compilation session directory `\\?\C:\Users\rasvi\Configura\web-rnd\cbb\target\debug\incremental\cbbapi-9d7mgtv9wzcv\s-g3in3aqwi6-fhge86-working`: The system cannot find the file 
specified. (os error 2)

error: internal compiler error: broken MIR in DefId(0:8434 ~ cbbapi[111c]::regional_state::content::{impl#0}::has_permission::{closure#0}) ((_3.0: global_state::group::GroupRef)): can't project out of PlaceTy { ty: regional_state::content::ContentScope, variant_index: None }
  --> cbbapi\src\regional_state\content.rs:41:96
   |
41 |       pub async fn has_permission(self, client: Client, global_state: &GlobalState) -> Result<()> {
   |  _________________________________________________________________________________________________^
42 | |         // This doesn't ICE
43 | |         // match self {
44 | |         //     ContentScope::Group(group_ref) => {
...  |
50 | |         client.check_permission(group_ref, global_state).await
51 | |     }
   | |_____^
   |
   = note: delayed at compiler\rustc_mir\src\borrow_check\type_check\mod.rs:308:27

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler\rustc_mir\src\borrow_check\type_check\mod.rs:777:20

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa\compiler\rustc_middle\src\ty\relate.rs:382:59

error: internal compiler error: broken MIR in DefId(0:8435 ~ cbbapi[111c]::regional_state::content::{impl#0}::has_permission::{closure#0}::{closure#0}) (((_1.0: global_state::group::GroupRef).0: global_state::group::GroupRef)): bad field access (global_state::customer::CustomerId: global_state::group::GroupRef): NoSolution
  --> cbbapi\src\regional_state\content.rs:49:33
   |
49 |         let ContentScope::Group(group_ref) = self;
   |                                 ^^^^^^^^^
   |
   = note: delayed at compiler\rustc_mir\src\borrow_check\type_check\mod.rs:308:27

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler\rustc_errors\src\lib.rs:1166:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0 (09c42c458 2021-10-18) running on x86_64-pc-windows-msvc

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by Noratrieb** [comment]
This issue does not have a reproduction and is therefore not really actionable, closing. If you find the problem again, please open a new issue, ideally with a reproduction.
